tensorflow/contrib/gdr/gdr_memory_manager.cc

```diff
@@ -0,0 +1,608 @@
+#ifdef TENSORFLOW_USE_GDR
+
+#include "tensorflow/contrib/gdr/gdr_memory_manager.h"
+
+#include <atomic>
+#include <cerrno>
+#include <fstream>
+#include <list>
+#include <map>
+
+#include <fcntl.h>
+#include <rdma/rdma_cma.h>
+#include <rdma/rdma_verbs.h>
+#include <sys/epoll.h>
+
+#include "tensorflow/contrib/gdr/gdr.pb.h"
+#include "tensorflow/core/common_runtime/bfc_allocator.h"
+#include "tensorflow/core/common_runtime/device.h"
+#include "tensorflow/core/common_runtime/dma_helper.h"
+#if GOOGLE_CUDA
+#include "tensorflow/core/common_runtime/gpu/gpu_util.h"
+#include "tensorflow/core/common_runtime/gpu/process_state.h"
+#endif  // GOOGLE_CUDA
+#include "tensorflow/core/framework/allocator_registry.h"
+#include "tensorflow/core/lib/core/status.h"
+#include "tensorflow/core/platform/macros.h"
+#include "tensorflow/core/platform/mutex.h"
+
+namespace tensorflow {
+
+namespace {
+
+bool IsGDRAvailable() {
+  std::ifstream ifs("/proc/modules");
```

Comment:
@shamoya Is GDR available on Windows? I am actually pretty interested in porting my patch to Win32, as there are already some APIs like [NetDMA](https://docs.microsoft.com/en-us/windows-hardware/drivers/network/netdma-drivers) or [Network Direct Kernel Provider Interface](https://docs.microsoft.com/en-us/windows-hardware/drivers/network/network-direct-kernel-programming-interface--ndkpi-). In a recent talk from MSR people, I heard that they could achieve pretty good improvement using RDMA to train RNN on CNTK, and I think it would be worth doing that on TF as well.